Adds moment card component

dhaval0894 opened the pull request:
Creates card component [#785](https://github.com/ifmeorg/ifme/issues/785)

julianguyen commented on client/app/bundles/momentCard/components/MomentCardViewers.jsx:
```diff
@@ -0,0 +1,20 @@
+  render() {
+    return (
+      <div>
+        Viewers
```
You should use the i18n key here

baohouse replied:
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

julianguyen commented on yarn-error.log:
```diff
@@ -0,0 +1,51 @@
+Arguments: 
+  /usr/bin/node /usr/bin/yarn lint:flow
```
We shouldn't be committing this file!

baohouse replied:
I added *.log to .gitignore, but you'll want to delete this file.

julianguyen commented on app/assets/stylesheets/application/shared.scss:
```diff
@@ -45,56 +47,74 @@
+  font-size: 10px;
+  letter-spacing: 0.1em;
+  text-transform: uppercase;
+  color: rgba(109, 8, 57, 0.5);
```
This color should be a variable in colors.scss 🌈

julianguyen commented:
You'll also want to rebase from master!

nshki commented:
Thanks a bunch for working on this @dhaval0894! Let me know once you've rebased with `master` and I'll take a look at everything in-depth.

baohouse commented on client/app/stories/index.jsx:
```diff
@@ -22,6 +22,7 @@ import DropdownFillSmall from '../bundles/shared/components/Dropdown/DropdownFil
 import Footer from '../bundles/shared/components/Footer/Footer';
```
When you rebase, this file should no longer exist. All the stories are now split up into separate files, so there should be one specifically for the MomentCards.

baohouse commented on client/app/bundles/momentCard/components/MomentCardSettings.jsx:
```diff
@@ -0,0 +1,32 @@
+};
+
+export default class MomentCardSettings extends
+  React.Component<MomentCardSettingsProp, MomentCardSettingsState> {
```
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

baohouse commented on client/app/bundles/momentCard/components/MomentCardName.jsx:
```diff
@@ -0,0 +1,23 @@
+};
+
+export default class MomentCardName extends
+  React.Component<MomentCardNameProp, MomentCardNameState> {
```
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

baohouse commented on client/app/bundles/momentCard/components/MomentCardMoods.jsx:
```diff
@@ -0,0 +1,25 @@
+};
+
+export default class MomentCardMoods
+  extends React.Component<MomentCardMoodsProp, MomentCardMoodsState> {
```
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

baohouse commented on client/app/bundles/momentCard/components/MomentCardDraft.jsx:
```diff
@@ -0,0 +1,23 @@
+  render() {
+    return (
+      <div className={css.draft}>
+        <Tag label={'DRAFT'} />
```
Allow the ability to change the wording of "Draft" to support multiple languages.

baohouse commented on client/app/bundles/momentCard/components/MomentCardDraft.jsx:
```diff
@@ -0,0 +1,23 @@
+};
+
+export default class MomentCardDraft
+  extends React.Component<MomentCardDraftProp, MomentCardDraftState> {
```
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

baohouse commented on client/app/bundles/momentCard/components/MomentCardDate.jsx:
```diff
@@ -0,0 +1,23 @@
+};
+
+export default class MomentCardDate
+  extends React.Component<MomentCardDateProp, MomentCardDateState> {
```
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

baohouse commented on client/app/bundles/momentCard/components/MomentCardCategories.jsx:
```diff
@@ -0,0 +1,25 @@
+};
+
+export default class MomentCardCategories
+  extends React.Component<MomentCardCategoriesProp, MomentCardCategoriesState> {
```
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

baohouse commented on client/app/bundles/momentCard/components/MomentCard.jsx:
```diff
@@ -0,0 +1,53 @@
+  cardType: string,
+};
+
+export default class MomentCard extends React.Component<MomentCardProp, MomentCardState> {
```
You can omit MomentCardViewersState as the second parameter (it's optional).

On the following lines, you can omit `props: MomentCardViewersProp` and `state: MomentCardViewersState` because they should already be defined as part of React.Component< MomentCardViewersProp>.

baohouse commented on client/app/bundles/momentCard/components/MomentCard.jsx:
```diff
@@ -0,0 +1,53 @@
+type MomentCardProp = {
+  item: {
+    name: string,
+    category?: [],
```
```
category?: Array<string>,
mood?: Array<string>
```

dhaval0894 commented:
Thanks a lot, @baohouse, @julianguyen, @nshki for taking time to review this. I have updated the PR with the requested changes, so please let me know if there are any further changes. Thanks!

baohouse commented:
<img width="423" alt="screen shot 2018-03-21 at 8 54 26 am" src="https://user-images.githubusercontent.com/4228012/37720902-83a6bdda-2ce5-11e8-9251-50289ab157f9.png">

I did a visual inspection.
1. There's an issue if the main text runs long; it collides with the icons.
2. The padding on the right should be increased; the icons are too close to the right edge.
3. Can we setup props that take in onClick handlers for each of the icons?

@nshki How should it behave if the lines run longer than 3 lines? Fixed height with hiding-clipping? Expand the height of the card?

baohouse commented:
@nshki I looked at the mockups. With regards to width, it looks like we should make this take up the width of the parent container to be used in grid systems.

nshki commented:
@baohouse Yup totally!

baohouse commented:
@dhaval0894 

1. So the card should be fluid; it should take up 100% of the width.
2. Moment cards should have a min-height, and be allowed to expand to fit content.
3. When displaying it in Storybook, you can put them inside a grid layout. My PR #864 will let us have access to Ant Design; you can look at my Logo story changes there to see how I'm adding a 2-column layout.

dhaval0894 commented:
@baohouse Cool, I'll make these changes.

dhaval0894 commented:
Hey @baohouse, I have made the following changes:
1) Made the card fluid. Now it will resize depending upon the content.
2) The main text won't collide with the icons. I have added the max-width and word-wrap property to ensure text gets wrapped if it's too long.
3) Added onClick handlers for each of the icons.
4) Wrapped card components in grid layout in the storybook (I had to install ant design package as the PR [#864](https://github.com/ifmeorg/ifme/pull/864) is not merged into master)
5)  Increased the padding on the right so that the icons won't be too close to the right edge.

Also, the viewers text was already there when I took over the PR from @julianguyen, I didn't know the rationale for that so I kept it :).

dhaval0894 commented:
![screenshot from 2018-03-23 17-32-58](https://user-images.githubusercontent.com/7266024/37836667-1165af6a-2ed9-11e8-8098-5c59715526b0.png)
![screenshot from 2018-03-23 17-33-09](https://user-images.githubusercontent.com/7266024/37836669-11c02080-2ed9-11e8-8f38-ac8d3c525ae7.png)
![screenshot from 2018-03-23 17-33-15](https://user-images.githubusercontent.com/7266024/37836671-11fea544-2ed9-11e8-8d89-56b8a162b2db.png)
@baohouse Just for your reference. Please review it and let me know if there are any further changes.

baohouse commented:
@nshki I see "Viewers" text below the icons. I didn't see this in the original mockups. Was this an ad hoc design change?

baohouse commented:
@dhaval0894 
1. Let's remove the viewers text.
2. Let's make the container `display: block` and take up the whole width of the parent container

@nshki Do we have a minimum height we want to aim for? I think if we should the height taken up by two-line title with tags and creation time.

nshki commented:
@baohouse Yup, that minimum height sounds good to me!

dhaval0894 commented:
@baohouse @nshki Do we really need the min-height? The card adjusts itself depending upon the content height.

dhaval0894 commented:
Yes, good point @baohouse. Setting min-height won't help because if the content height is greater than the min-height then also cards will have different height. Like you said, allowing the implementor to override the component CSS to establish a min-height is a good way to do this :)

dhaval0894 commented:
@nshki @baohouse I have made these changes:
1) Remove the viewer's text.
2) Make the container display: block and take up the whole width of the parent container.
Let me know if there are any changes.

baohouse commented:
@dhaval0894 Mostly looks good. Only thing is that upgrading the parser gem broken `bundle install`. Can you update it to `parser (2.5.0.5)`?

dhaval0894 commented:
Hey @baohouse, updated it to `parser(2.5.0.5)`.

baohouse commented:
@julianguyen I looked at your comments, they've all been addressed. Feel free to approve your requested changes.
